HW06
================

Steps:

1.  Read in the 45 processed RNA-Seq data found in
    “./RNA\_Seq\_processed”
2.  Map gene.ids to gene.symbols (which I have for you)
3.  For 27 genes of interest AND your favorite gene, perform a t-test to
    see if it is differentially expressed between the WT vs H3K27M
    samples
4.  Create a graphing function and then create a boxplot that graphs
    expression between the two groups

#### Code

``` r
#create list of files contained by each directory 
H3K27M_list <- list.files("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/H3K27M")
WT_list <- list.files("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/WT")
```

``` r
library(readr)

#create empty dataset that loop results will go into 
H3K27M_df <- data.frame()
WT_df <- data.frame()

setwd("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/H3K27M")
#loop to read in mutated data
for (file in 1:length(H3K27M_list)) {
  H3K27M_data <- read_tsv(H3K27M_list[file])
  H3K27M_df <- rbind(H3K27M_df, H3K27M_data)
}
```

    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )

``` r
setwd("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/WT")
#loop to read in WT data
for (file in 1:length(WT_list)) {
  WT_data <- read_tsv(WT_list[file])
    WT_df <- rbind(WT_df, WT_data)
}
```

    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )
    ## Parsed with column specification:
    ## cols(
    ##   Name = col_character(),
    ##   Length = col_double(),
    ##   EffectiveLength = col_double(),
    ##   TPM = col_double(),
    ##   NumReads = col_double()
    ## )

``` r
library(tidyverse)
```

    ## ── Attaching packages ───────────────────────────────────────────────────── tidyverse 1.3.0 ──

    ## ✓ ggplot2 3.3.2.9000     ✓ dplyr   1.0.1     
    ## ✓ tibble  3.0.3          ✓ stringr 1.4.0     
    ## ✓ tidyr   1.1.1          ✓ forcats 0.5.0     
    ## ✓ purrr   0.3.4

    ## ── Conflicts ──────────────────────────────────────────────────────── tidyverse_conflicts() ──
    ## x dplyr::filter() masks stats::filter()
    ## x dplyr::lag()    masks stats::lag()

``` r
library(magrittr)
```

    ## 
    ## Attaching package: 'magrittr'

    ## The following object is masked from 'package:purrr':
    ## 
    ##     set_names

    ## The following object is masked from 'package:tidyr':
    ## 
    ##     extract

``` r
library(dplyr)

#add type column to distinguish WT from mutated
WT_df <- WT_df %>% mutate("type" = "WT")
H3K27M_df <- H3K27M_df %>% mutate("type" = "H3K27_Mutated")
```

``` r
#join into single dataframe for subsequent analysis
#my data is in long format, which is not what Robert's code was intended for, so see note below on some of Robert's code I had to omit 
GBM.transcripts <- WT_df %>%
full_join(H3K27M_df) %>%
  dplyr::rename("gene_id"="Name")
```

    ## Joining, by = c("Name", "Length", "EffectiveLength", "TPM", "NumReads", "type")

Using Robert’s code

``` r
#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#BiocManager::install("EnsDb.Hsapiens.v75")
library(ensembldb)
```

    ## Loading required package: BiocGenerics

    ## Loading required package: parallel

    ## 
    ## Attaching package: 'BiocGenerics'

    ## The following objects are masked from 'package:parallel':
    ## 
    ##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    ##     clusterExport, clusterMap, parApply, parCapply, parLapply,
    ##     parLapplyLB, parRapply, parSapply, parSapplyLB

    ## The following objects are masked from 'package:dplyr':
    ## 
    ##     combine, intersect, setdiff, union

    ## The following objects are masked from 'package:stats':
    ## 
    ##     IQR, mad, sd, var, xtabs

    ## The following objects are masked from 'package:base':
    ## 
    ##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    ##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    ##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    ##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    ##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    ##     union, unique, unsplit, which, which.max, which.min

    ## Loading required package: GenomicRanges

    ## Loading required package: stats4

    ## Loading required package: S4Vectors

    ## 
    ## Attaching package: 'S4Vectors'

    ## The following objects are masked from 'package:dplyr':
    ## 
    ##     first, rename

    ## The following object is masked from 'package:tidyr':
    ## 
    ##     expand

    ## The following object is masked from 'package:base':
    ## 
    ##     expand.grid

    ## Loading required package: IRanges

    ## 
    ## Attaching package: 'IRanges'

    ## The following objects are masked from 'package:dplyr':
    ## 
    ##     collapse, desc, slice

    ## The following object is masked from 'package:purrr':
    ## 
    ##     reduce

    ## Loading required package: GenomeInfoDb

    ## Loading required package: GenomicFeatures

    ## Loading required package: AnnotationDbi

    ## Loading required package: Biobase

    ## Welcome to Bioconductor
    ## 
    ##     Vignettes contain introductory material; view with
    ##     'browseVignettes()'. To cite Bioconductor, see
    ##     'citation("Biobase")', and for packages 'citation("pkgname")'.

    ## 
    ## Attaching package: 'AnnotationDbi'

    ## The following object is masked from 'package:dplyr':
    ## 
    ##     select

    ## Loading required package: AnnotationFilter

    ## 
    ## Attaching package: 'AnnotationFilter'

    ## The following object is masked from 'package:magrittr':
    ## 
    ##     not

    ## 
    ## Attaching package: 'ensembldb'

    ## The following object is masked from 'package:dplyr':
    ## 
    ##     filter

    ## The following object is masked from 'package:stats':
    ## 
    ##     filter

``` r
library(EnsDb.Hsapiens.v75)


ens.GBM.transcripts <- GBM.transcripts %>% 
mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

#Because I combined my data long-wise, many values would be removed with the following code (because my gene names show up more than once). Robert told me that it's okay to not use the following functions and still be able to do my analysis. 

#ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 

ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values
```

Do the t-test and make a table of the t-test results\!

``` r
#detach functions gave me an error, so I just specified which library some of my functions are from  
  #detach(package:EnsDb.Hsapiens.v75, unload = TRUE)
  #detach(package:ensembldb, unload = TRUE)
library(broom)

#added CDC42 as my own gene of interest
genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "PTEN", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2", "CDC42")


GBM.genes.of.interest <- dplyr::filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)

#save table so I don't have to rerun all my previous code to do my analysis
#write_csv(GBM.genes.of.interest, "GBM.genes.of.interest.csv")

GBM.genes.of.interest <- read.csv("GBM.genes.of.interest.csv")


#creating function to filter by gene and then group by type
grouped_genes <- function(x){
  GBM.genes.of.interest %>%  
  dplyr::filter(gene_symbol == x) %>%
    group_by(type)
}

Group_genes <- lapply(genes_of_interest, grouped_genes)

#creating function to conduct two-sample t-test
ttest <- function(x){
  t.test(TPM~type, data = x) }
  
ttest_values <- lapply(Group_genes,ttest)

#relabel so it is clear what gene the statistics belong to
for (x in 1:length(ttest_values))  {
ttest_values[[x]]$data.name <-genes_of_interest[[x]]
 }

#print out the t-test results
print(ttest_values)
```

    ## [[1]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  IRX1
    ## t = 5.1963, df = 38.468, p-value = 6.964e-06
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  1.347856 3.067221
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   3.2049706                   0.9974324 
    ## 
    ## 
    ## [[2]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  OSR1
    ## t = 7.0746, df = 82.332, p-value = 4.52e-10
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  0.4618776 0.8232156
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                  0.73168325                  0.08913667 
    ## 
    ## 
    ## [[3]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  DCHS2
    ## t = 4.5305, df = 131, p-value = 1.31e-05
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  0.05659998 0.14434186
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                  0.12384557                  0.02337465 
    ## 
    ## 
    ## [[4]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  BRINP3
    ## t = 5.7003, df = 81.377, p-value = 1.856e-07
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  2.420455 5.015942
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   4.6269138                   0.9087153 
    ## 
    ## 
    ## [[5]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  TOB2P1
    ## t = -4.4518, df = 28.477, p-value = 0.00012
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.25810948 -0.09551828
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                  0.07228035                  0.24909423 
    ## 
    ## 
    ## [[6]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  FOXD1
    ## t = -4.4471, df = 22.83, p-value = 0.0001875
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.7547856 -0.2753868
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.0544923                   0.5695785 
    ## 
    ## 
    ## [[7]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  ZFPM2
    ## t = 5.9804, df = 232.09, p-value = 8.345e-09
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  0.2453977 0.4865292
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.5138162                   0.1478527 
    ## 
    ## 
    ## [[8]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  GLB1
    ## t = -3.7167, df = 780.82, p-value = 0.0002162
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.4439066 -0.1370630
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.5345477                   0.8250325 
    ## 
    ## 
    ## [[9]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  ALG5
    ## t = -2.6011, df = 155.85, p-value = 0.01019
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.8158627 -0.1115663
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.5033586                   0.9670731 
    ## 
    ## 
    ## [[10]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  TRIM4
    ## t = -3.3098, df = 183.01, p-value = 0.001125
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.6815307 -0.1724539
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.6766466                   1.1036389 
    ## 
    ## 
    ## [[11]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  ADARB2
    ## t = 5.4005, df = 216.48, p-value = 1.743e-07
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  0.07071557 0.15199675
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                  0.15032038                  0.03896422 
    ## 
    ## 
    ## [[12]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  PCDHGA11
    ## t = -1.5016, df = 43.792, p-value = 0.1404
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -1.3477572  0.1969575
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.1440745                   0.7194743 
    ## 
    ## 
    ## [[13]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  IDH1
    ## t = -1.1107, df = 332.7, p-value = 0.2675
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -1.338811  0.372509
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    1.468825                    1.951976 
    ## 
    ## 
    ## [[14]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  EGFR
    ## t = -1.8237, df = 232.78, p-value = 0.06948
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -3.6778217  0.1420498
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.8026329                   2.5705189 
    ## 
    ## 
    ## [[15]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  MGMT
    ## t = 0.69646, df = 176.33, p-value = 0.4871
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.08528337  0.17830575
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.2516160                   0.2051048 
    ## 
    ## 
    ## [[16]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  TERT
    ## t = -1.0798, df = 215.44, p-value = 0.2814
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.019736907  0.005765887
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                 0.007112426                 0.014097936 
    ## 
    ## 
    ## [[17]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  PTEN
    ## t = -2.7256, df = 179.14, p-value = 0.007056
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -1.8245065 -0.2921004
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    2.663606                    3.721909 
    ## 
    ## 
    ## [[18]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  TP53
    ## t = -0.99726, df = 722.11, p-value = 0.319
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.5533164  0.1805442
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.5484932                   0.7348793 
    ## 
    ## 
    ## [[19]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  RB1
    ## t = -2.1264, df = 197.67, p-value = 0.03471
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.72175575 -0.02717904
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    0.691275                    1.065742 
    ## 
    ## 
    ## [[20]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  PTEN
    ## t = -2.7256, df = 179.14, p-value = 0.007056
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -1.8245065 -0.2921004
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    2.663606                    3.721909 
    ## 
    ## 
    ## [[21]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  ATRX
    ## t = -2.2052, df = 267.22, p-value = 0.02829
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.48980193 -0.02773258
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.7070009                   0.9657682 
    ## 
    ## 
    ## [[22]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  PDGFRA
    ## t = 1.0505, df = 275.38, p-value = 0.2944
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -1.787307  5.877505
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    4.813524                    2.768425 
    ## 
    ## 
    ## [[23]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  PIK3CA
    ## t = -0.43859, df = 175.55, p-value = 0.6615
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.4892553  0.3113365
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    1.098380                    1.187339 
    ## 
    ## 
    ## [[24]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  MICA
    ## t = -2.2467, df = 43.204, p-value = 0.02982
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -3.2774150 -0.1770228
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.1675271                   1.8947460 
    ## 
    ## 
    ## [[25]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  CDKN2A
    ## t = -2.604, df = 308.33, p-value = 0.00966
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -1.034988 -0.144060
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                  0.06456416                  0.65408801 
    ## 
    ## 
    ## [[26]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  EZH2
    ## t = -1.0244, df = 441.11, p-value = 0.3062
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.23196548  0.07300789
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.3972694                   0.4767482 
    ## 
    ## 
    ## [[27]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  BRD2
    ## t = -2.9219, df = 758.05, p-value = 0.003582
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -0.6737638 -0.1322481
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                   0.9597985                   1.3628045 
    ## 
    ## 
    ## [[28]]
    ## 
    ##  Welch Two Sample t-test
    ## 
    ## data:  CDC42
    ## t = -2.3486, df = 176.27, p-value = 0.01996
    ## alternative hypothesis: true difference in means is not equal to 0
    ## 95 percent confidence interval:
    ##  -4.9703679 -0.4312928
    ## sample estimates:
    ## mean in group H3K27_Mutated            mean in group WT 
    ##                    4.834198                    7.535028

Now create a graphing function to create boxplots to visualize the
results. Plot expression on the y-axis. The graph should look like [this
example](IRX1_example.png)

``` r
#selecting columns of interest to make new dataframe
trimmed_df <- GBM.genes.of.interest %>%  
  dplyr::select(gene_symbol, type, TPM)

#create graphing function
graph_tpm <- function(df) {
      plot <- ggplot(subset(trimmed_df, trimmed_df$gene_symbol == genes_of_interest[i]),
              aes(type, TPM, fill=type)) +
      geom_boxplot() +
      theme_bw()+
      scale_x_discrete(name = genes_of_interest[i]) +
      scale_y_continuous(name = "Expression_(TPM)") +
      ggtitle(paste(genes_of_interest[i], "Expression in GBM models \nby H3K27 Mutated or WT Status")) +
      theme(plot.title = element_text(hjust = 0.5)) +
      theme(legend.position = "none") 
  print(plot)
}

#using for loop with the graphing function
for(i in 1: length(genes_of_interest)) {
  graph_tpm(trimmed_df) 
}
```

![](HW06_files/figure-gfm/create%20a%20graphing%20function-1.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-2.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-3.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-4.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-5.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-6.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-7.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-8.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-9.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-10.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-11.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-12.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-13.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-14.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-15.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-16.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-17.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-18.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-19.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-20.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-21.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-22.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-23.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-24.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-25.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-26.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-27.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-28.png)<!-- -->

``` r
sessionInfo()
```

    ## R version 4.0.2 (2020-06-22)
    ## Platform: x86_64-apple-darwin17.0 (64-bit)
    ## Running under: macOS High Sierra 10.13.6
    ## 
    ## Matrix products: default
    ## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
    ## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
    ## 
    ## locale:
    ## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
    ## 
    ## attached base packages:
    ## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
    ## [8] methods   base     
    ## 
    ## other attached packages:
    ##  [1] broom_0.7.0               EnsDb.Hsapiens.v75_2.99.0
    ##  [3] ensembldb_2.12.1          AnnotationFilter_1.12.0  
    ##  [5] GenomicFeatures_1.40.1    AnnotationDbi_1.50.3     
    ##  [7] Biobase_2.48.0            GenomicRanges_1.40.0     
    ##  [9] GenomeInfoDb_1.24.2       IRanges_2.22.2           
    ## [11] S4Vectors_0.26.1          BiocGenerics_0.34.0      
    ## [13] magrittr_1.5              forcats_0.5.0            
    ## [15] stringr_1.4.0             dplyr_1.0.1              
    ## [17] purrr_0.3.4               tidyr_1.1.1              
    ## [19] tibble_3.0.3              ggplot2_3.3.2.9000       
    ## [21] tidyverse_1.3.0           readr_1.3.1              
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] ProtGenerics_1.20.0         matrixStats_0.56.0         
    ##  [3] bitops_1.0-6                fs_1.5.0                   
    ##  [5] lubridate_1.7.9             bit64_4.0.2                
    ##  [7] progress_1.2.2              httr_1.4.2                 
    ##  [9] tools_4.0.2                 backports_1.1.8            
    ## [11] R6_2.4.1                    lazyeval_0.2.2             
    ## [13] DBI_1.1.0                   colorspace_1.4-1           
    ## [15] withr_2.2.0                 tidyselect_1.1.0           
    ## [17] prettyunits_1.1.1           bit_4.0.4                  
    ## [19] curl_4.3                    compiler_4.0.2             
    ## [21] cli_2.0.2                   rvest_0.3.6                
    ## [23] xml2_1.3.2                  DelayedArray_0.14.1        
    ## [25] labeling_0.3                rtracklayer_1.48.0         
    ## [27] scales_1.1.1                askpass_1.1                
    ## [29] rappdirs_0.3.1              digest_0.6.25              
    ## [31] Rsamtools_2.4.0             rmarkdown_2.3              
    ## [33] XVector_0.28.0              pkgconfig_2.0.3            
    ## [35] htmltools_0.5.0             dbplyr_1.4.4               
    ## [37] rlang_0.4.7                 readxl_1.3.1               
    ## [39] rstudioapi_0.11             RSQLite_2.2.0              
    ## [41] farver_2.0.3                generics_0.0.2             
    ## [43] jsonlite_1.7.0              BiocParallel_1.22.0        
    ## [45] RCurl_1.98-1.2              GenomeInfoDbData_1.2.3     
    ## [47] Matrix_1.2-18               Rcpp_1.0.5                 
    ## [49] munsell_0.5.0               fansi_0.4.1                
    ## [51] lifecycle_0.2.0             stringi_1.4.6              
    ## [53] yaml_2.2.1                  SummarizedExperiment_1.18.2
    ## [55] zlibbioc_1.34.0             BiocFileCache_1.12.1       
    ## [57] grid_4.0.2                  blob_1.2.1                 
    ## [59] crayon_1.3.4                lattice_0.20-41            
    ## [61] Biostrings_2.56.0           haven_2.3.1                
    ## [63] hms_0.5.3                   knitr_1.29                 
    ## [65] pillar_1.4.6                biomaRt_2.44.1             
    ## [67] reprex_0.3.0                XML_3.99-0.5               
    ## [69] glue_1.4.1                  evaluate_0.14              
    ## [71] modelr_0.1.8                vctrs_0.3.2                
    ## [73] cellranger_1.1.0            gtable_0.3.0               
    ## [75] openssl_1.4.2               assertthat_0.2.1           
    ## [77] xfun_0.16                   GenomicAlignments_1.24.0   
    ## [79] memoise_1.1.0               ellipsis_0.3.1
