HW06
================
Laura Cespedes Arias (with lots of help from the team)

In cancer, mutations arise that promote growth or survival of cells. In
glioblastoma multiforme and other high grade gliomas, a common mutation
is a mutation of the 27th lysine (K) to a methionine (M) of the histone
subunit H3, or in short H3K27M.

H3K27M is the most frequent oncohistone in brain cancers, but the
biology is still not well understood. Your analysis is to look at the
expression of several (27) genes to see if they are differentially
expressed and plot 27 boxplots each gene. The data used in this analysis
was obtained [from this
publication](https://pubmed.ncbi.nlm.nih.gov/31085178/)

Steps:

1.  Read in the 45 processed RNA-Seq data found in
    “./RNA\_Seq\_processed”
2.  Map gene.ids to gene.symbols (which I have for you)
3.  For 27 genes of interest AND your favorite gene, perform a t-test to
    see if it is differentially expressed between the WT vs H3K27M
    samples
4.  Create a graphing function and then create a boxplot that graphs
    expression between the two groups

#### Code

From the RNA-Seq files, you only need the “Name” from one file and the
“TPM” column from all the files. TPM stands for “transcripts per
million” and is a common unit for normalized expression data.

``` r
library(tidyverse) 
```

    ## Warning: package 'tibble' was built under R version 3.6.2

    ## Warning: package 'purrr' was built under R version 3.6.2

``` r
library(dplyr) 
library(knitr)

#First, get the list of all file names (and then paste "./RNA_Seq_processed" to get the full path)
file_names_WT <- list.files("./RNA_Seq_processed/WT", full.names=TRUE)
file_names_H3K27M <- list.files("./RNA_Seq_processed/H3K27M", full.names=TRUE)

#Read in all files! After reading files, use function select within an apply to get tables with the columns of interest. Then, change the TPM column name to include info on the WT/H3K27M status

#I realize creating a function might have been much better, but I could not get them to work! Therefore, I decided to use for loops and apply functions which I have used more often in the past :)

# WT
files_WT_full <- lapply(file_names_WT, read_tsv)
files_WT <- lapply(files_WT_full, select,contains(c("Name","TPM")))
for (i in 1: length(files_WT))
{
  colnames(files_WT[[i]]) <- c("gene_id",paste0("TPM_",i,"_WT")) 
}

# H3K27M
files_H3K27M_full <- lapply(file_names_H3K27M, read_tsv)
files_H3K27M <- lapply(files_H3K27M_full, select,contains(c("Name","TPM")))
for (i in 1: length(files_H3K27M))
{
  colnames(files_H3K27M[[i]]) <- c("gene_id",paste0("TPM_",i,"_H3K27M")) 
}

#Get a list with all files!
files <- c(files_WT,files_H3K27M)

#Merge all the files in one table. I used a for loop and the function "full join" to acomplish this.
GBM.transcripts <- files[[1]]
for (i in 2: length(files))
{
  GBM.transcripts <- full_join(GBM.transcripts,files[[i]],by = "gene_id") 
}
```

Now, install the packages commented below (if needed), and then use this
code to map the transcript IDs to gene symbols. To use this code, you
need a dataframe called GBM.transcripts that has the first column
“gene\_id” that contains the transcript ids (e.g. ENST00000456328.2)
and the remaining columns contain the TPM data. So long as the first
column contains the “gene\_id” column as mentioned above, this *should*
run.

``` r
#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#NN
#install.packages("XML", repos = "http://www.omegahat.net/R")
library(XML)
library(ensembldb)
library(EnsDb.Hsapiens.v75)

ens.GBM.transcripts <- GBM.transcripts %>% 
  mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 
ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values
```

Do the t-test and make a table of the t-test
results\!

``` r
#run this code to unload the libraries from before, it might be helpful because the select() function from dplyr might be hidden otherwise
detach(package:EnsDb.Hsapiens.v75, unload = T)
detach(package:ensembldb, unload = T)

#add in your own gene of interest!!! 
genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "PTEN", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2")

GBM.genes.of.interest <- filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)
GBM.genes.of.interest$gene_symbol
```

    ##  [1] "BRINP3"   "OSR1"     "IDH1"     "GLB1"     "PIK3CA"   "PDGFRA"  
    ##  [7] "DCHS2"    "TERT"     "IRX1"     "FOXD1"    "PCDHGA11" "TOB2P1"  
    ## [13] "MICA"     "BRD2"     "EGFR"     "TRIM4"    "EZH2"     "ZFPM2"   
    ## [19] "CDKN2A"   "ADARB2"   "PTEN"     "MGMT"     "ALG5"     "RB1"     
    ## [25] "TP53"     "ATRX"

``` r
#Now perform a t-test between the H3K mutated and the wt samples. There are many ways to do this actually, you can use a for loop or you could do the tidy alternative with broom(), but the for loop is probably the easiest

# I got the idea to use "gather" and then "spread" from my teammate Ricardo who kindly showed me his code after I got stuck :). 

# Tidy table so the genes are columns insteads of rows:
# For WT data
WT_table <- GBM.genes.of.interest %>%
  dplyr::select(gene_symbol, ends_with("_WT")) %>%
  gather(key = file, value = value, -gene_symbol) %>%
  spread(key = gene_symbol, value = value)

# For H3K27M data
H3K27M_table <- GBM.genes.of.interest %>%
  dplyr::select(gene_symbol, ends_with("_H3K27M")) %>%
  gather(key = "file", value = "value", -gene_symbol) %>%
  spread(key = gene_symbol, value = value)

#Now, to perform the t test, create empty vectors to store p values and t values
Genes <- colnames(WT_table)
p_value <- c()
t_stat <- c()

#Then, with a for loop, iterate across all columns (i.e. genes), starting from column 2 (column 1 = "file"). 
for(i in 2:ncol(WT_table))
{
  p_value[i] <- t.test(WT_table[,i],H3K27M_table[,i])$p.value
  t_stat[i] <- t.test(WT_table[,i],H3K27M_table[,i])$statistic
}

#Lastly, combine the vectors to create a data table with gene_id, p value and t statistic values
t_results <- data.frame(Genes,p_value,t_stat)[-1,]

#print out the t-test results
knitr::kable(t_results, format = "markdown", col.names = c("Gene", "p-value","t statistic"))
```

|    | Gene     |   p-value | t statistic |
| :- | :------- | --------: | ----------: |
| 2  | ADARB2   | 0.0000020 | \-5.9637598 |
| 3  | ALG5     | 0.0000817 |   4.5112646 |
| 4  | ATRX     | 0.3072447 |   1.0340966 |
| 5  | BRD2     | 0.1385084 | \-1.5165130 |
| 6  | BRINP3   | 0.0000306 | \-4.8544677 |
| 7  | CDKN2A   | 0.0430789 |   2.1521262 |
| 8  | DCHS2    | 0.0000023 | \-5.9011566 |
| 9  | EGFR     | 0.1642918 |   1.4387841 |
| 10 | EZH2     | 0.9641300 |   0.0452742 |
| 11 | FOXD1    | 0.0001875 |   4.4471325 |
| 12 | GLB1     | 0.0003784 |   3.9652748 |
| 13 | IDH1     | 0.1403354 |   1.5131411 |
| 14 | IRX1     | 0.0000070 | \-5.1962706 |
| 15 | MGMT     | 0.3558852 | \-0.9332835 |
| 16 | MICA     | 0.0288078 |   2.3464674 |
| 17 | OSR1     | 0.0000013 | \-5.9497259 |
| 18 | PCDHGA11 | 0.2512366 |   1.1795213 |
| 19 | PDGFRA   | 0.7507788 |   0.3209236 |
| 20 | PIK3CA   | 0.7354268 |   0.3401311 |
| 21 | PTEN     | 0.0235358 |   2.3630214 |
| 22 | RB1      | 0.1202909 |   1.6092247 |
| 23 | TERT     | 0.6531105 |   0.4528648 |
| 24 | TOB2P1   | 0.0001200 |   4.4518241 |
| 25 | TP53     | 0.8962146 |   0.1312187 |
| 26 | TRIM4    | 0.0003872 |   3.9212887 |
| 27 | ZFPM2    | 0.0001880 | \-4.3256464 |

Now create a graphing function to create boxplots to visualize the
results. Plot expression on the y-axis. The graph should look like [this
example](IRX1_example.png)

``` r
#to work in the tidyverse, it will be easier to make tidy the dataframe first

#What I did first, was to create a new column in both H3K27M and WT tables. This column will have the information on mutation status, which will be useful when both tables will be merged (to be able to later use as categorical variable for the graphing function)
H3K27M_table$Class <-as.factor("H3K27M")
WT_table$Class <-as.factor("WT")

#Merge tables
table <- bind_rows(H3K27M_table,WT_table)
```

    ## Warning in bind_rows_(x, .id): Unequal factor levels: coercing to character

    ## Warning in bind_rows_(x, .id): binding character and factor vector, coercing
    ## into character vector
    
    ## Warning in bind_rows_(x, .id): binding character and factor vector, coercing
    ## into character vector

``` r
#Graphing function
#For this function, it is only necessary to specify the gene_id (columns in data frame, see below). Then is a simple function to create a boxplot similar to the one in the example. I used aes_string instead of aes so it can identify 'gene_id' in each iteration. The categorical variable (x axis and color) is "Class" the variable that I created at the beggining of the chunk.

figures_FUN <- function(gene_id = "column_name"){
  ggplot(table, aes_string(x = "Class", y = gene_id, fill = "Class"))+
    geom_boxplot() +
    labs(x = "H3K27M mutation" ,y = "Expression (TPM)",title=paste(gene_id, "Expression in GBM models by H3K27 mutated or WT status")) +
    theme_bw() +
    theme(legend.position = "none") 
  }

#then use a for loop combined with the graphing function to make a graph for all your genes of interest 

#To be able to use the function within a for loop it is necessary to have a vector with gene_ids (this corresponds to the column names of the table, excluding the "file" and "Clas" variables)
genes <- colnames(table[,-c(1,28)])

#Then, to create a new directory where plots will be stored
dir.create("plots") 
```

    ## Warning in dir.create("plots"): 'plots' already exists

``` r
#For loop to save .png files with the figures for all genes (to the directory created above)
for(j in 1:length(genes))
{
  figures_FUN(gene_id = genes[j])
  ggsave(filename=paste0(genes[j],"_plot.png"),plot = last_plot(),device="png",path="./plots")
}
```

    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image
    ## Saving 7 x 5 in image

``` r
sessionInfo()
```

    ## R version 3.6.1 (2019-07-05)
    ## Platform: x86_64-apple-darwin15.6.0 (64-bit)
    ## Running under: macOS High Sierra 10.13.6
    ## 
    ## Matrix products: default
    ## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
    ## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
    ## 
    ## locale:
    ## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
    ## 
    ## attached base packages:
    ## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
    ## [8] methods   base     
    ## 
    ## other attached packages:
    ##  [1] AnnotationFilter_1.10.0 GenomicFeatures_1.38.2  AnnotationDbi_1.48.0   
    ##  [4] Biobase_2.46.0          GenomicRanges_1.38.0    GenomeInfoDb_1.22.1    
    ##  [7] IRanges_2.20.2          S4Vectors_0.24.4        BiocGenerics_0.32.0    
    ## [10] XML_3.99-0              knitr_1.26              forcats_0.4.0          
    ## [13] stringr_1.4.0           dplyr_0.8.5             purrr_0.3.4            
    ## [16] readr_1.3.1             tidyr_1.0.2             tibble_3.0.1           
    ## [19] ggplot2_3.3.0           tidyverse_1.3.0        
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] nlme_3.1-140                ProtGenerics_1.18.0        
    ##  [3] matrixStats_0.56.0          bitops_1.0-6               
    ##  [5] fs_1.3.1                    lubridate_1.7.4            
    ##  [7] bit64_4.0.2                 progress_1.2.2             
    ##  [9] httr_1.4.1                  tools_3.6.1                
    ## [11] backports_1.1.6             R6_2.4.1                   
    ## [13] lazyeval_0.2.2              DBI_1.1.0                  
    ## [15] colorspace_1.4-1            withr_2.2.0                
    ## [17] tidyselect_1.0.0            prettyunits_1.1.1          
    ## [19] bit_4.0.4                   curl_4.3                   
    ## [21] compiler_3.6.1              cli_2.0.2                  
    ## [23] rvest_0.3.5                 xml2_1.2.2                 
    ## [25] DelayedArray_0.12.3         labeling_0.3               
    ## [27] rtracklayer_1.46.0          scales_1.1.0               
    ## [29] askpass_1.1                 rappdirs_0.3.1             
    ## [31] Rsamtools_2.2.3             digest_0.6.25              
    ## [33] rmarkdown_2.0               XVector_0.26.0             
    ## [35] pkgconfig_2.0.3             htmltools_0.4.0            
    ## [37] highr_0.8                   dbplyr_1.4.2               
    ## [39] rlang_0.4.5                 readxl_1.3.1               
    ## [41] rstudioapi_0.11             RSQLite_2.2.0              
    ## [43] farver_2.0.3                generics_0.0.2             
    ## [45] jsonlite_1.6.1              BiocParallel_1.20.1        
    ## [47] RCurl_1.98-1.2              magrittr_1.5               
    ## [49] GenomeInfoDbData_1.2.2      Matrix_1.2-17              
    ## [51] Rcpp_1.0.4.6                munsell_0.5.0              
    ## [53] fansi_0.4.1                 lifecycle_0.2.0            
    ## [55] stringi_1.4.6               yaml_2.2.0                 
    ## [57] SummarizedExperiment_1.16.1 zlibbioc_1.32.0            
    ## [59] BiocFileCache_1.10.2        grid_3.6.1                 
    ## [61] blob_1.2.1                  crayon_1.3.4               
    ## [63] lattice_0.20-38             Biostrings_2.54.0          
    ## [65] haven_2.2.0                 hms_0.5.2                  
    ## [67] pillar_1.4.3                biomaRt_2.42.1             
    ## [69] reprex_0.3.0                glue_1.4.0                 
    ## [71] evaluate_0.14               modelr_0.1.5               
    ## [73] vctrs_0.2.4                 cellranger_1.1.0           
    ## [75] gtable_0.3.0                openssl_1.4.1              
    ## [77] assertthat_0.2.1            xfun_0.11                  
    ## [79] broom_0.5.3                 GenomicAlignments_1.22.1   
    ## [81] memoise_1.1.0               ellipsis_0.3.0
