HW06
================

In cancer, mutations arise that promote growth or survival of cells. In
glioblastoma multiforme and other high grade gliomas, a common mutation
is a mutation of the 27th lysine (K) to a methionine (M) of the histone
subunit H3, or in short H3K27M.

H3K27M is the most frequent oncohistone in brain cancers, but the
biology is still not well understood. Your analysis is to look at the
expression of several (27) genes to see if they are differentially
expressed and plot 27 boxplots each gene. The data used in this analysis
was obtained [from this
publication](https://pubmed.ncbi.nlm.nih.gov/31085178/)

Steps:

1.  Read in the 45 processed RNA-Seq data found in
    “./RNA\_Seq\_processed”
2.  Map gene.ids to gene.symbols (which I have for you)
3.  For 27 genes of interest AND your favorite gene, perform a t-test to
    see if it is differentially expressed between the WT vs H3K27M
    samples
4.  Create a graphing function and then create a boxplot that graphs
    expression between the two groups

#### Code

From the RNA-Seq files, you only need the “Name” from one file and the
“TPM” column from all the files. TPM stands for “transcripts per
million” and is a common unit for normalized expression data.

``` r
library("readr")
library("dplyr")
```

    ## 
    ## Attaching package: 'dplyr'

    ## The following objects are masked from 'package:stats':
    ## 
    ##     filter, lag

    ## The following objects are masked from 'package:base':
    ## 
    ##     intersect, setdiff, setequal, union

``` r
library("tidyr")
library("ggplot2")
#hint, using apply (specifically sapply) you can read in the data into a list and then bind the columns together. Or you can use a for loop too. 

H3K27M_filepath_list <- list.files(path = "RNA_Seq_processed/H3K27M", full.names = TRUE)
WT_filepath_list <- list.files(path = "RNA_Seq_processed/WT", full.names = TRUE) #list.files() simply creates a list containing the file names of all the files in a folder. With the full.names argument set to TRUE, the function outputs the path of the file instead of just the file name

H3K27M_filename_list <- unlist(list.files(path = "RNA_Seq_processed/H3K27M"))
WT_filename_list <- unlist(list.files(path = "RNA_Seq_processed/WT")) #and this creates a list of the file names without the path, and converts to a character vector using unlist()

H3K27M_filename_list <- gsub(".sf", replacement = "_H3K", H3K27M_filename_list)
WT_filename_list <- gsub(".sf", replacement = "_WT", WT_filename_list) #updates these variables to filenames without extensions
```

``` r
H3K_RNA_list <- list()
WT_RNA_list <- list()

for(file in 1:length(H3K27M_filepath_list)) { #I might be able to use an apply function here instead, might try later
  H3K_RNA_list[[file]] <- read_tsv(H3K27M_filepath_list[[file]]) #the NAME column is a character vector - we will see if this matters
}
for(file in 1:length(WT_filepath_list)) {
  WT_RNA_list[[file]] <- read_tsv(WT_filepath_list[[file]])
} #THis code reads in the files for H3K27M and WT data into a list of data frames
  
for(file in 1:length(H3K_RNA_list)) { #I can likely select for only TPM columns using contain() once the for loop is complete - I will do that later if I have the time
  H3K_RNA_list[[file]] <- H3K_RNA_list[[file]] %>% 
  select(Name, TPM)
}
for(file in 1:length(WT_RNA_list)) {
  WT_RNA_list[[file]] <- WT_RNA_list[[file]] %>% 
  select(Name, TPM) #This code selects for the Name and TPM column
}

H3K_RNA_Name <- H3K_RNA_list[[1]]
H3K_RNA_join <- H3K_RNA_Name[1]

WT_RNA_Name <- WT_RNA_list[[1]]
WT_RNA_join <- WT_RNA_Name[1] #This code creates a data frame where the first and only column is the name column of the data
```

``` r
for(df in 1:length(H3K_RNA_list)) {
  H3K_RNA_join <- inner_join(H3K_RNA_list[[df]], H3K_RNA_join, by = "Name")
}
for(df in 1:length(WT_RNA_list)) {
  WT_RNA_join <- inner_join(WT_RNA_list[[df]], WT_RNA_join, by = "Name")
}

rev_H3K_filename <- rev(H3K27M_filename_list) #This reverse the order of the object because the for() loop adds the columns in a reverse order
colnames(H3K_RNA_join) <- c("gene_id", rev_H3K_filename) #This works!

rev_WT_filename <- rev(WT_filename_list)
colnames(WT_RNA_join) <- c("gene_id", rev_WT_filename)


GBM.transcripts <- H3K_RNA_join %>% 
  inner_join(WT_RNA_join, by = "gene_id")
```

Now, install the packages commented below (if needed), and then use this
code to map the transcript IDs to gene symbols. To use this code, you
need a dataframe called GBM.transcripts that has the first column
“gene\_id” that contains the transcript ids (e.g. ENST00000456328.2)
and the remaining columns contain the TPM data. So long as the first
column contains the “gene\_id” column as mentioned above, this *should*
run.

``` r
#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#BiocManager::install("EnsDb.Hsapiens.v75")
library(ensembldb)
library(EnsDb.Hsapiens.v75)

ens.GBM.transcripts <- GBM.transcripts %>% 
  mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 
ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values
```

Do the t-test and make a table of the t-test results\!

``` r
#run this code to unload the libraries from before, it might be helpful because the select() function from dplyr might be hidden otherwise
detach(package:EnsDb.Hsapiens.v75, unload = T)
detach(package:ensembldb, unload = T)

#add in your own gene of interest!!! 
genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2", "PAM")
#gene "PTEN" is repeated here, so I got rid of the extra
#I picked the gene PAM, since I studied the protein during undergrad
GBM.genes.of.interest <- filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)
genes_of_interest_2 <- as.factor(rev(GBM.genes.of.interest$gene_symbol))

#Now perform a t-test between the H3K mutated and the wt samples. There are many ways to do this actually, you can use a for loop or you could do the tidy alternative with broom(), but the for loop is probably the easiest
library("broom")
t_test <- data.frame()
for(gene in 1:length(genes_of_interest)) {
t_test <- tidy(t.test(x = GBM.genes.of.interest[gene, 2:24],
                      y = GBM.genes.of.interest[gene, 25:46])) %>% 
  bind_rows(t_test)  
}
t_test <- mutate(t_test, gene_symbol = genes_of_interest_2) %>% 
  dplyr::select(gene_symbol, everything())
#print out the t-test results

print(t_test)
```

    ## # A tibble: 27 x 11
    ##    gene_symbol estimate estimate1 estimate2 statistic p.value parameter conf.low
    ##    <fct>          <dbl>     <dbl>     <dbl>     <dbl>   <dbl>     <dbl>    <dbl>
    ##  1 ATRX        -0.136      0.371     0.507    -1.03   3.07e-1      40.4  -0.402 
    ##  2 TP53        -0.00171    0.0260    0.0277   -0.131  8.96e-1      43.0  -0.0281
    ##  3 RB1         -1.07       2.33      3.41     -1.61   1.20e-1      24.7  -2.45  
    ##  4 ALG5        -2.05       1.80      3.86     -4.51   8.17e-5      32.0  -2.98  
    ##  5 MGMT         0.173      0.862     0.689     0.933  3.56e-1      43.0  -0.201 
    ##  6 PTEN        -0.969      2.06      3.03     -2.36   2.35e-2      36.7  -1.80  
    ##  7 ADARB2       0.434      0.514     0.0801    5.96   2.01e-6      28.0   0.285 
    ##  8 CDKN2A      -0.876      0.0505    0.926    -2.15   4.31e-2      21.2  -1.72  
    ##  9 ZFPM2        0.718      0.970     0.252     4.33   1.88e-4      26.8   0.378 
    ## 10 EZH2        -0.00975    0.646     0.655    -0.0453 9.64e-1      37.3  -0.446 
    ## # ... with 17 more rows, and 3 more variables: conf.high <dbl>, method <chr>,
    ## #   alternative <chr>

Now create a graphing function to create boxplots to visualize the
results. Plot expression on the y-axis. The graph should look like [this
example](IRX1_example.png)

**I had difficulty from this point on, so I’ll wait to see how others
cleaned their data and graphed**

``` r
#to work in the tidyverse, it will be easier to make tidy the dataframe first
t_test_2 <- t_test %>% 
  dplyr::select(gene_symbol, estimate1, estimate2, conf.low, conf.high) %>%  
  gather(key = "cell_line_type", value = "mean", estimate1:estimate2) %>% 
  group_by(gene_symbol) # group_by() doesn't seem to be working here


#create a graphing function

graphing_function <- function(gene) {
  ggplot(aes(x = cell_line_type, y = ))
}
#then use a for loop combined with the graphing function to make a graph for all your genes of interest 

t_test_2 %>% 
  filter(gene_symbol == "ATRX") %>% 
  ggplot(aes(x = cell_line_type, y = mean)) +
  geom_boxplot(aes(lower = conf.low, upper = conf.high))
```

![](HW06_files/figure-gfm/create%20a%20graphing%20function-1.png)<!-- -->

``` r
sessionInfo()
```

    ## R version 4.0.2 (2020-06-22)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 10 x64 (build 18362)
    ## 
    ## Matrix products: default
    ## 
    ## locale:
    ## [1] LC_COLLATE=English_United States.1252 
    ## [2] LC_CTYPE=English_United States.1252   
    ## [3] LC_MONETARY=English_United States.1252
    ## [4] LC_NUMERIC=C                          
    ## [5] LC_TIME=English_United States.1252    
    ## 
    ## attached base packages:
    ## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
    ## [8] methods   base     
    ## 
    ## other attached packages:
    ##  [1] broom_0.7.0             AnnotationFilter_1.12.0 GenomicFeatures_1.40.1 
    ##  [4] AnnotationDbi_1.50.3    Biobase_2.48.0          GenomicRanges_1.40.0   
    ##  [7] GenomeInfoDb_1.24.2     IRanges_2.22.2          S4Vectors_0.26.1       
    ## [10] BiocGenerics_0.34.0     ggplot2_3.3.2           tidyr_1.1.1            
    ## [13] dplyr_1.0.1             readr_1.3.1            
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] httr_1.4.2                  bit64_4.0.2                
    ##  [3] assertthat_0.2.1            askpass_1.1                
    ##  [5] BiocFileCache_1.12.1        blob_1.2.1                 
    ##  [7] GenomeInfoDbData_1.2.3      Rsamtools_2.4.0            
    ##  [9] yaml_2.2.1                  progress_1.2.2             
    ## [11] pillar_1.4.6                RSQLite_2.2.0              
    ## [13] backports_1.1.7             lattice_0.20-41            
    ## [15] glue_1.4.1                  digest_0.6.25              
    ## [17] XVector_0.28.0              colorspace_1.4-1           
    ## [19] htmltools_0.5.0             Matrix_1.2-18              
    ## [21] XML_3.99-0.5                pkgconfig_2.0.3            
    ## [23] biomaRt_2.44.1              zlibbioc_1.34.0            
    ## [25] purrr_0.3.4                 scales_1.1.1               
    ## [27] BiocParallel_1.22.0         tibble_3.0.3               
    ## [29] openssl_1.4.2               farver_2.0.3               
    ## [31] generics_0.0.2              ellipsis_0.3.1             
    ## [33] withr_2.2.0                 SummarizedExperiment_1.18.2
    ## [35] lazyeval_0.2.2              cli_2.0.2                  
    ## [37] magrittr_1.5                crayon_1.3.4               
    ## [39] memoise_1.1.0               evaluate_0.14              
    ## [41] fansi_0.4.1                 tools_4.0.2                
    ## [43] prettyunits_1.1.1           hms_0.5.3                  
    ## [45] lifecycle_0.2.0             matrixStats_0.56.0         
    ## [47] stringr_1.4.0               munsell_0.5.0              
    ## [49] DelayedArray_0.14.1         Biostrings_2.56.0          
    ## [51] compiler_4.0.2              rlang_0.4.7                
    ## [53] grid_4.0.2                  RCurl_1.98-1.2             
    ## [55] rappdirs_0.3.1              labeling_0.3               
    ## [57] bitops_1.0-6                rmarkdown_2.3              
    ## [59] gtable_0.3.0                DBI_1.1.0                  
    ## [61] curl_4.3                    R6_2.4.1                   
    ## [63] GenomicAlignments_1.24.0    knitr_1.29                 
    ## [65] rtracklayer_1.48.0          utf8_1.1.4                 
    ## [67] bit_4.0.4                   ProtGenerics_1.20.0        
    ## [69] stringi_1.4.6               Rcpp_1.0.5                 
    ## [71] vctrs_0.3.2                 dbplyr_1.4.4               
    ## [73] tidyselect_1.1.0            xfun_0.16
