HW06
================
Daniel Krochmal

In cancer, mutations arise that promote growth or survival of cells. In
glioblastoma multiforme and other high grade gliomas, a common mutation
is a mutation of the 27th lysine (K) to a methionine (M) of the histone
subunit H3, or in short H3K27M.

H3K27M is the most frequent oncohistone in brain cancers, but the
biology is still not well understood. Your analysis is to look at the
expression of several (27) genes to see if they are differentially
expressed and plot 27 boxplots each gene. The data used in this analysis
was obtained [from this
publication](https://pubmed.ncbi.nlm.nih.gov/31085178/)

Steps:

1.  Read in the 45 processed RNA-Seq data found in
    “./RNA\_Seq\_processed”
2.  Map gene.ids to gene.symbols (which I have for you)
3.  For 27 genes of interest AND your favorite gene, perform a t-test to
    see if it is differentially expressed between the WT vs H3K27M
    samples
4.  Create a graphing function and then create a boxplot that graphs
    expression between the two groups

#### Code

From the RNA-Seq files, you only need the “Name” from one file and the
“TPM” column from all the files. TPM stands for “transcripts per
million” and is a common unit for normalized expression data.

``` r
#Defining factors mainly for better control over the order of boxplots later on
type <- factor(c("H3K27M_mutated", "WT"))

#This chunk creates a list of .sf files in specified subfolder, batch loads them in, binds them together and applies "H3K27M_mutated" or "WT" factor to the values
H3K27M_list <- Sys.glob("RNA_Seq_processed/H3K27M/*.sf")
H3K27M_data <- lapply(H3K27M_list, function(x){read_tsv(x)})
H3K27M.df <- do.call("rbind", H3K27M_data) %>%
  select(Name, TPM) %>%
  mutate(type = "H3K27M_mutated") %>%
  rename(gene_id = Name)

WT_list <- Sys.glob("RNA_Seq_processed/WT/*.sf")
WT_data <- lapply(WT_list, function(x){read_tsv(x)})
WT.df <- do.call("rbind", WT_data) %>%
  select(Name, TPM) %>%
  mutate(type = "WT") %>%
  rename(gene_id = Name)

#Binding WT and H3K27M dataframes together and generating input for Robert's script
GBM.transcripts <- WT.df %>%
  rbind(H3K27M.df)
```

Now, install the packages commented below (if needed), and then use this
code to map the transcript IDs to gene symbols. To use this code, you
need a dataframe called GBM.transcripts that has the first column
“gene\_id” that contains the transcript ids (e.g. ENST00000456328.2)
and the remaining columns contain the TPM data. So long as the first
column contains the “gene\_id” column as mentioned above, this *should*
run.

``` r
#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#BiocManager::install("EnsDb.Hsapiens.v75")
library(ensembldb)
library(EnsDb.Hsapiens.v75)

ens.GBM.transcripts <- GBM.transcripts %>% 
  mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

#Since I prepared my data frame long-wise and I intended to have duplicated gene symbols (each duplicate has their own TPM values assigned), I'm commenting out the line below

#ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 
ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values
```

``` r
detach(package:EnsDb.Hsapiens.v75, unload = T)
detach(package:ensembldb, unload = T)
```

Do the t-test and make a table of the t-test results\!

``` r
genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "PTEN", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2", "RN7SL2") #added RN7SL2

GBM.genes.of.interest <- filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)

for (gene in genes_of_interest) {
  
  WT_gene <- GBM.genes.of.interest %>%
    filter(gene_symbol == gene, type == "WT")
  
  mut_gene <- GBM.genes.of.interest %>%
    filter(gene_symbol == gene, type == "H3K27M_mutated")
  
  #I simplified the output to show only the p-values
  pvalue <- t.test(mut_gene$TPM, WT_gene$TPM)$p.value
  print(paste0(gene, " ", signif(pvalue, digits = 3)))
}
```

    ## [1] "IRX1 6.96e-06"
    ## [1] "OSR1 4.52e-10"
    ## [1] "DCHS2 1.31e-05"
    ## [1] "BRINP3 1.86e-07"
    ## [1] "TOB2P1 0.00012"
    ## [1] "FOXD1 0.000188"
    ## [1] "ZFPM2 8.35e-09"
    ## [1] "GLB1 0.000216"
    ## [1] "ALG5 0.0102"
    ## [1] "TRIM4 0.00112"
    ## [1] "ADARB2 1.74e-07"
    ## [1] "PCDHGA11 0.14"
    ## [1] "IDH1 0.267"
    ## [1] "EGFR 0.0695"
    ## [1] "MGMT 0.487"
    ## [1] "TERT 0.281"
    ## [1] "PTEN 0.00706"
    ## [1] "TP53 0.319"
    ## [1] "RB1 0.0347"
    ## [1] "PTEN 0.00706"
    ## [1] "ATRX 0.0283"
    ## [1] "PDGFRA 0.294"
    ## [1] "PIK3CA 0.661"
    ## [1] "MICA 0.0298"
    ## [1] "CDKN2A 0.00966"
    ## [1] "EZH2 0.306"
    ## [1] "BRD2 0.00358"
    ## [1] "RN7SL2 0.303"

Now create a graphing function to create boxplots to visualize the
results. Plot expression on the y-axis. The graph should look like [this
example](IRX1_example.png)

``` r
library(ggplot2)

#Creating a graphing function
graphIt <- function(gene) {
  
  gene_sele <- GBM.genes.of.interest %>%
    filter(gene_symbol == gene)
  
    plot <- ggplot(gene_sele, aes(x = type, y = TPM, fill = type)) +
      geom_boxplot() +
      
      theme_bw() +
      
      theme(
        legend.position = "none",
        plot.title = element_text(hjust = 0.5)
      ) +
      
      labs(
        x = "H3K27M",
        y = "Expression_(TPM)",
        title = paste0(gene, " Expression in GBM models\nby H3K27M Mutated or WT Status")
      )
    
    print(plot)
}

#Creating a for loop that calls the graphing function over all genes in genes_of_interest vector
for (gene in genes_of_interest) {
  graphIt(gene)
}
```

![](HW06_files/figure-gfm/create%20a%20graphing%20function-1.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-2.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-3.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-4.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-5.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-6.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-7.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-8.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-9.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-10.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-11.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-12.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-13.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-14.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-15.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-16.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-17.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-18.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-19.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-20.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-21.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-22.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-23.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-24.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-25.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-26.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-27.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-28.png)<!-- -->

``` r
sessionInfo()
```

    ## R version 4.0.2 (2020-06-22)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 10 x64 (build 18362)
    ## 
    ## Matrix products: default
    ## 
    ## locale:
    ## [1] LC_COLLATE=English_United States.1252 
    ## [2] LC_CTYPE=English_United States.1252   
    ## [3] LC_MONETARY=English_United States.1252
    ## [4] LC_NUMERIC=C                          
    ## [5] LC_TIME=English_United States.1252    
    ## 
    ## attached base packages:
    ## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
    ## [8] methods   base     
    ## 
    ## other attached packages:
    ##  [1] AnnotationFilter_1.12.0 GenomicFeatures_1.40.1  AnnotationDbi_1.50.3   
    ##  [4] Biobase_2.48.0          GenomicRanges_1.40.0    GenomeInfoDb_1.24.2    
    ##  [7] IRanges_2.22.2          S4Vectors_0.26.1        BiocGenerics_0.34.0    
    ## [10] forcats_0.5.0           stringr_1.4.0           purrr_0.3.4            
    ## [13] tidyr_1.1.1             tibble_3.0.2            ggplot2_3.3.2          
    ## [16] tidyverse_1.3.0         knitr_1.29              dplyr_1.0.0            
    ## [19] readr_1.3.1            
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] ProtGenerics_1.20.0         matrixStats_0.56.0         
    ##  [3] bitops_1.0-6                fs_1.5.0                   
    ##  [5] lubridate_1.7.9             bit64_4.0.2                
    ##  [7] progress_1.2.2              httr_1.4.2                 
    ##  [9] tools_4.0.2                 backports_1.1.7            
    ## [11] R6_2.4.1                    lazyeval_0.2.2             
    ## [13] DBI_1.1.0                   colorspace_1.4-1           
    ## [15] withr_2.2.0                 tidyselect_1.1.0           
    ## [17] prettyunits_1.1.1           bit_4.0.4                  
    ## [19] curl_4.3                    compiler_4.0.2             
    ## [21] cli_2.0.2                   rvest_0.3.6                
    ## [23] xml2_1.3.2                  DelayedArray_0.14.1        
    ## [25] labeling_0.3                rtracklayer_1.48.0         
    ## [27] scales_1.1.1                askpass_1.1                
    ## [29] rappdirs_0.3.1              digest_0.6.25              
    ## [31] Rsamtools_2.4.0             rmarkdown_2.3              
    ## [33] XVector_0.28.0              pkgconfig_2.0.3            
    ## [35] htmltools_0.5.0             dbplyr_1.4.4               
    ## [37] rlang_0.4.6                 readxl_1.3.1               
    ## [39] rstudioapi_0.11             RSQLite_2.2.0              
    ## [41] farver_2.0.3                generics_0.0.2             
    ## [43] jsonlite_1.7.0              BiocParallel_1.22.0        
    ## [45] RCurl_1.98-1.2              magrittr_1.5               
    ## [47] GenomeInfoDbData_1.2.3      Matrix_1.2-18              
    ## [49] Rcpp_1.0.4.6                munsell_0.5.0              
    ## [51] fansi_0.4.1                 lifecycle_0.2.0            
    ## [53] stringi_1.4.6               yaml_2.2.1                 
    ## [55] SummarizedExperiment_1.18.2 zlibbioc_1.34.0            
    ## [57] BiocFileCache_1.12.1        grid_4.0.2                 
    ## [59] blob_1.2.1                  crayon_1.3.4               
    ## [61] lattice_0.20-41             Biostrings_2.56.0          
    ## [63] haven_2.3.1                 hms_0.5.3                  
    ## [65] pillar_1.4.6                biomaRt_2.44.1             
    ## [67] reprex_0.3.0                XML_3.99-0.5               
    ## [69] glue_1.4.1                  evaluate_0.14              
    ## [71] modelr_0.1.8                vctrs_0.3.1                
    ## [73] cellranger_1.1.0            gtable_0.3.0               
    ## [75] openssl_1.4.2               assertthat_0.2.1           
    ## [77] xfun_0.15                   broom_0.7.0                
    ## [79] GenomicAlignments_1.24.0    memoise_1.1.0              
    ## [81] ellipsis_0.3.1
