HW06
================

In cancer, mutations arise that promote growth or survival of cells. In
glioblastoma multiforme and other high grade gliomas, a common mutation
is a mutation of the 27th lysine (K) to a methionine (M) of the histone
subunit H3, or in short H3K27M.

H3K27M is the most frequent oncohistone in brain cancers, but the
biology is still not well understood. Your analysis is to look at the
expression of several (27) genes to see if they are differentially
expressed and plot 27 boxplots each gene. The data used in this analysis
was obtained [from this
publication](https://pubmed.ncbi.nlm.nih.gov/31085178/)

Steps:

1.  Read in the 45 processed RNA-Seq data found in
    “./RNA\_Seq\_processed”
2.  Map gene.ids to gene.symbols (which I have for you)
3.  For 27 genes of interest AND your favorite gene, perform a t-test to
    see if it is differentially expressed between the WT vs H3K27M
    samples
4.  Create a graphing function and then create a boxplot that graphs
    expression between the two groups

#### Code

From the RNA-Seq files, you only need the “Name” from one file and the
“TPM” column from all the files. TPM stands for “transcripts per
million” and is a common unit for normalized expression data.

``` r
#hint, using apply (specifically sapply) you can read in the data into a list and then bind the columns together. Or you can use a for loop too. 

#you only need the 

library(tidyverse)

#To establish a vector with the all the filepaths to read
H3K27M_filepath <- list.files("RNA_Seq_processed/H3K27M", full.names = TRUE) #for the H3K27 folder
WT_filepath <- list.files("RNA_Seq_processed/WT", full.names = TRUE) #for the WT folder
RNA_seq_file <- c(H3K27M_filepath, WT_filepath) #To get a complete vector of all the filepaths

#To establish a vector with just the file names to keep track of where the data came from
H3K27M_names <- list.files("RNA_Seq_processed/H3K27M")
WT_names <- list.files("RNA_Seq_processed/WT")
#And also to be able to track which ones where from the WT and H3K27 folders
H3K27M_names <- gsub(pattern = "\\.sf", H3K27M_names, replacement = "/H3")
WT_names <- gsub(pattern = "\\.sf", WT_names, replacement = "/WT")
column_names <- c("gene_id", H3K27M_names, WT_names) #And with this we have the complete list of names for the GBM.transcripts matrix


#Function for importing and cleaning the data
import_clean <- function(x){
  if(x == "RNA_Seq_processed/H3K27M/BT245_C19.sf") {
    read_tsv(x) %>%
      select(Name, TPM) %>%
      rename(gene_id = Name) #Since we need only one column with the Name, only one file will keep that column...
  } else {
    read_tsv(x) %>%
    select(TPM) #...and the rest only get the TPM column
}
  }
  

H3K27M <- sapply(RNA_seq_file, import_clean) #apply the new function to every filepath required into a list

GBM.transcripts <- bind_cols(H3K27M) #Now we simply stashed every element of the list into to columns of a data frame

colnames(GBM.transcripts) <- column_names #and with this we rename the columns to the required names. gene_id for the Name column, and the file name where the TPM column came from, as well as the folder (WT/H3K27)
```

Now, install the packages commented below (if needed), and then use this
code to map the transcript IDs to gene symbols. To use this code, you
need a dataframe called GBM.transcripts that has the first column
“gene\_id” that contains the transcript ids (e.g. ENST00000456328.2)
and the remaining columns contain the TPM data. So long as the first
column contains the “gene\_id” column as mentioned above, this *should*
run.

``` r
#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#BiocManager::install("EnsDb.Hsapiens.v75")
library(ensembldb)
library(EnsDb.Hsapiens.v75)

ens.GBM.transcripts <- GBM.transcripts %>% 
  mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 
ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values
```

Do the t-test and make a table of the t-test results\!

``` r
#run this code to unload the libraries from before, it might be helpful because the select() function from dplyr might be hidden otherwise
detach(package:EnsDb.Hsapiens.v75, unload = T)
detach(package:ensembldb, unload = T)

#add in your own gene of interest!!! 
genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "PTEN", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2", "CICP27") #I chose CICP27


GBM.genes.of.interest <- filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)

#Now perform a t-test between the H3K mutated and the wt samples. There are many ways to do this actually, you can use a for loop or you could do the tidy alternative with broom(), but the for loop is probably the easiest

#print out the t-test results

#Since we can track the original folder, we can create two separate data frames and perform paired t-test between them.
#I also found more intuitive to flip the data frames to have the gene names as columns
WT <- GBM.genes.of.interest %>%
  dplyr::select(gene_symbol, ends_with("WT")) %>%
  gather(key = "file", value = "value", 2:23) %>%
  spread(key = gene_symbol, value = value)


H3 <- GBM.genes.of.interest %>%
  dplyr::select(gene_symbol, ends_with("H3")) %>%
  gather(key = "file", value = "value", 2:24) %>%
  spread(key = gene_symbol, value = value)

gene_list <- colnames(WT) #We need a vector with all the gene names. Careful, this will include the "file" column, as well. 

tests <- list() #Empty list to work with
tests_table <- matrix(nrow = 27, ncol = 3, data = NA) #Empty data frame to work with

for(i in 2:ncol(H3)) { 
  tests[[gene_list[i]]] <- t.test(H3[,i],WT[,i]) #Add the result of every t-test to a list
  gene <- gene_list[i] #Add the name of every gene to a vector
  statistic <- round(tests[[i-1]][[1]], digits = 4) #Select only the statistics value and add them to a vector
  p.value <- round(tests[[i-1]][[3]], digits = 6) #Select only the p-value and add them to a vector
    tests_table[i-1,] <- c(gene, statistic, p.value) #Add gene, statistic, and p.value to the empty data frame in order
}

knitr::kable(tests_table, format = "markdown", col.names = c("Gene", "t-statistic", "p-value")) #Print the table with the results
```

| Gene     | t-statistic | p-value  |
| :------- | :---------- | :------- |
| ADARB2   | 5.9638      | 2e-06    |
| ALG5     | \-4.5113    | 8.2e-05  |
| ATRX     | \-1.0341    | 0.307245 |
| BRD2     | 1.5165      | 0.138508 |
| BRINP3   | 4.8545      | 3.1e-05  |
| CDKN2A   | \-2.1521    | 0.043079 |
| CICP27   | 0.9727      | 0.341228 |
| DCHS2    | 5.9012      | 2e-06    |
| EGFR     | \-1.4388    | 0.164292 |
| EZH2     | \-0.0453    | 0.96413  |
| FOXD1    | \-4.4471    | 0.000188 |
| GLB1     | \-3.9653    | 0.000378 |
| IDH1     | \-1.5131    | 0.140335 |
| IRX1     | 5.1963      | 7e-06    |
| MGMT     | 0.9333      | 0.355885 |
| MICA     | \-2.3465    | 0.028808 |
| OSR1     | 5.9497      | 1e-06    |
| PCDHGA11 | \-1.1795    | 0.251237 |
| PDGFRA   | \-0.3209    | 0.750779 |
| PIK3CA   | \-0.3401    | 0.735427 |
| PTEN     | \-2.363     | 0.023536 |
| RB1      | \-1.6092    | 0.120291 |
| TERT     | \-0.4529    | 0.653111 |
| TOB2P1   | \-4.4518    | 0.00012  |
| TP53     | \-0.1312    | 0.896215 |
| TRIM4    | \-3.9213    | 0.000387 |
| ZFPM2    | 4.3256      | 0.000188 |

Now create a graphing function to create boxplots to visualize the
results. Plot expression on the y-axis. The graph should look like [this
example](IRX1_example.png)

``` r
#to work in the tidyverse, it will be easier to make tidy the dataframe first

#For this last part, I flipped the entire GBM.genes.of.interest data frame and the I create a new parameter that considers if the data came from the WT or the H3K27 data frame, which is going to be useful to add a color aesthetic the the plots.
GBM.tidied <- GBM.genes.of.interest %>%
  gather(key = "file", value = "value", 2:46) %>%
  separate(file, sep = "/", into = c("file", "H3K27")) %>%
  spread(key = gene_symbol, value = value)

GBM.tidied$H3K27[GBM.tidied$H3K27 == "H3"] <- "H3K27_Mutated" #In order to have the labels as requested

theme_set(theme_bw()) #General theme of the plot


#create a graphing function

graph_function <- function(gene = "name") {
  ggplot(GBM.tidied, aes_string(x = "H3K27", y = gene, fill = "H3K27")) + #It seems that aes() doesn't work well by itself, so I chose aes_string() to input thing in "quotes".
  geom_boxplot() +
  labs(y = "Expression_(TPM)",
       title = paste(gene, "Expression in GBM models \nby H3K27 Mutated or WT Status")) + #This way it adds the corresponding gene name to each plot
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5))
}


#then use a for loop combined with the graphing function to make a graph for all your genes of interest 

gene_names <- colnames(GBM.tidied) #We need a vector with the gene names
gene_names <- gene_names[3:29] #To only have the gene names in quotation marks


for(t in 1:length(gene_names)){
  print(graph_function(gene = gene_names[t])) #This way we can visualize de plots in the md file
}
```

![](HW06_files/figure-gfm/create%20a%20graphing%20function-1.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-2.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-3.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-4.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-5.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-6.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-7.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-8.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-9.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-10.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-11.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-12.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-13.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-14.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-15.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-16.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-17.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-18.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-19.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-20.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-21.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-22.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-23.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-24.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-25.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-26.png)<!-- -->![](HW06_files/figure-gfm/create%20a%20graphing%20function-27.png)<!-- -->

``` r
#Enjoy the automatic creation of all 27 plots! :D
```

``` r
sessionInfo()
```

    ## R version 4.0.1 (2020-06-06)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 10 x64 (build 18362)
    ## 
    ## Matrix products: default
    ## 
    ## locale:
    ## [1] LC_COLLATE=Spanish_Mexico.1252  LC_CTYPE=Spanish_Mexico.1252   
    ## [3] LC_MONETARY=Spanish_Mexico.1252 LC_NUMERIC=C                   
    ## [5] LC_TIME=Spanish_Mexico.1252    
    ## 
    ## attached base packages:
    ## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
    ## [8] methods   base     
    ## 
    ## other attached packages:
    ##  [1] AnnotationFilter_1.12.0 GenomicFeatures_1.40.1  AnnotationDbi_1.50.3   
    ##  [4] Biobase_2.48.0          GenomicRanges_1.40.0    GenomeInfoDb_1.24.2    
    ##  [7] IRanges_2.22.2          S4Vectors_0.26.1        BiocGenerics_0.34.0    
    ## [10] forcats_0.5.0           stringr_1.4.0           dplyr_1.0.0            
    ## [13] purrr_0.3.4             readr_1.3.1             tidyr_1.1.1            
    ## [16] tibble_3.0.1            ggplot2_3.3.2           tidyverse_1.3.0        
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] ProtGenerics_1.20.0         matrixStats_0.56.0         
    ##  [3] bitops_1.0-6                fs_1.5.0                   
    ##  [5] lubridate_1.7.9             bit64_4.0.2                
    ##  [7] progress_1.2.2              httr_1.4.2                 
    ##  [9] tools_4.0.1                 backports_1.1.7            
    ## [11] R6_2.4.1                    lazyeval_0.2.2             
    ## [13] DBI_1.1.0                   colorspace_1.4-1           
    ## [15] withr_2.2.0                 tidyselect_1.1.0           
    ## [17] prettyunits_1.1.1           bit_4.0.4                  
    ## [19] curl_4.3                    compiler_4.0.1             
    ## [21] cli_2.0.2                   rvest_0.3.6                
    ## [23] xml2_1.3.2                  DelayedArray_0.14.1        
    ## [25] labeling_0.3                rtracklayer_1.48.0         
    ## [27] scales_1.1.1                askpass_1.1                
    ## [29] rappdirs_0.3.1              digest_0.6.25              
    ## [31] Rsamtools_2.4.0             rmarkdown_2.3              
    ## [33] XVector_0.28.0              pkgconfig_2.0.3            
    ## [35] htmltools_0.5.0             highr_0.8                  
    ## [37] dbplyr_1.4.4                rlang_0.4.6                
    ## [39] readxl_1.3.1                rstudioapi_0.11            
    ## [41] RSQLite_2.2.0               farver_2.0.3               
    ## [43] generics_0.0.2              jsonlite_1.7.0             
    ## [45] BiocParallel_1.22.0         RCurl_1.98-1.2             
    ## [47] magrittr_1.5                GenomeInfoDbData_1.2.3     
    ## [49] Matrix_1.2-18               Rcpp_1.0.4.6               
    ## [51] munsell_0.5.0               fansi_0.4.1                
    ## [53] lifecycle_0.2.0             stringi_1.4.6              
    ## [55] yaml_2.2.1                  SummarizedExperiment_1.18.2
    ## [57] zlibbioc_1.34.0             BiocFileCache_1.12.1       
    ## [59] grid_4.0.1                  blob_1.2.1                 
    ## [61] crayon_1.3.4                lattice_0.20-41            
    ## [63] Biostrings_2.56.0           haven_2.3.1                
    ## [65] hms_0.5.3                   knitr_1.29                 
    ## [67] pillar_1.4.6                biomaRt_2.44.1             
    ## [69] reprex_0.3.0                XML_3.99-0.5               
    ## [71] glue_1.4.1                  evaluate_0.14              
    ## [73] modelr_0.1.8                vctrs_0.3.1                
    ## [75] cellranger_1.1.0            gtable_0.3.0               
    ## [77] openssl_1.4.1               assertthat_0.2.1           
    ## [79] xfun_0.15                   broom_0.7.0                
    ## [81] GenomicAlignments_1.24.0    memoise_1.1.0              
    ## [83] ellipsis_0.3.1
