HW06
================

In cancer, mutations arise that promote growth or survival of cells. In
glioblastoma multiforme and other high grade gliomas, a common mutation
is a mutation of the 27th lysine (K) to a methionine (M) of the histone
subunit H3, or in short H3K27M.

H3K27M is the most frequent oncohistone in brain cancers, but the
biology is still not well understood. Your analysis is to look at the
expression of several (27) genes to see if they are differentially
expressed and plot 27 boxplots each gene. The data used in this analysis
was obtained [from this
publication](https://pubmed.ncbi.nlm.nih.gov/31085178/)

Steps:

1.  Read in the 45 processed RNA-Seq data found in
    “./RNA\_Seq\_processed”
2.  Map gene.ids to gene.symbols (which I have for you)
3.  For 27 genes of interest AND your favorite gene, perform a t-test to
    see if it is differentially expressed between the WT vs H3K27M
    samples
4.  Create a graphing function and then create a boxplot that graphs
    expression between the two groups

#### Code

From the RNA-Seq files, you only need the “Name” from one file and the
“TPM” column from all the files. TPM stands for “transcripts per
million” and is a common unit for normalized expression data.

``` r
library(tidyverse)
library(knitr)


#hint, using apply (specifically sapply) you can read in the data into a list and then bind the columns together. Or you can use a for loop too. 
#you only need the 

# use list.files to gather all the files in the RNA_Seq_processed folder
h3k27m_files <- list.files("RNA_Seq_processed/H3K27M", full.names = TRUE)
wt_files <- list.files("RNA_Seq_processed/WT", full.names = TRUE)
# combine the wt and h3k27m files
all_files <- c(h3k27m_files, wt_files)


# generate a function that uses an if{} statement to read in the appropriate columns from each file
read_function <- function(x) { 
  if(x == all_files[1]) { 
    read_tsv(x) %>% 
      dplyr::select(Name, TPM)
  }
  else { 
    read_tsv(x) %>% 
     dplyr:: select(TPM) 
  }
}

# apply the function to all files in the RNA_Seq_processed folder
all_files2 <- sapply(all_files, read_function)

# bind the columns and save as GBM.transcripts
GBM.transcripts <- bind_cols(all_files2)

# rename the Name column to gene_id
names(GBM.transcripts)[names(GBM.transcripts) == "Name"] <- "gene_id"
```

Now, install the packages commented below (if needed), and then use this
code to map the transcript IDs to gene symbols. To use this code, you
need a dataframe called GBM.transcripts that has the first column
“gene\_id” that contains the transcript ids (e.g. ENST00000456328.2)
and the remaining columns contain the TPM data. So long as the first
column contains the “gene\_id” column as mentioned above, this *should*
run.

``` r
#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#BiocManager::install("EnsDb.Hsapiens.v75")

library(ensembldb)
```

    ## Warning: package 'GenomicFeatures' was built under R version 4.0.2

    ## Warning: package 'AnnotationDbi' was built under R version 4.0.2

``` r
library(EnsDb.Hsapiens.v75)

ens.GBM.transcripts <- GBM.transcripts %>% 
  mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 
ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values
```

Do the t-test and make a table of the t-test
results\!

``` r
#run this code to unload the libraries from before, it might be helpful because the select() function from dplyr might be hidden otherwise
detach(package:EnsDb.Hsapiens.v75, unload = T)
detach(package:ensembldb, unload = T)

#add in your own gene of interest!!! 
# I added STK38, DIS3L2, and PIEZO1

genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "PTEN", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2", "STK38", "DIS3L2", "PIEZO1")


GBM.genes.of.interest <- filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)

#Now perform a t-test between the H3K mutated and the wt samples. There are many ways to do this actually, you can use a for loop or you could do the tidy alternative with broom(), but the for loop is probably the easiest

# generate new dataframe with only the H3K27M data and tidy
H3K27M <- GBM.genes.of.interest %>%
  dplyr::select(gene_symbol, TPM...2:TPM...24) %>% # select the columns that correspond 
  gather(key = "TPM_number", value = "value", 2:24, -gene_symbol) %>% # gather TPM values into a column
  spread(key = gene_symbol, value = value) # spread TPM values by gene_symbol


# do the same for WT data
WT <- GBM.genes.of.interest %>%
  dplyr::select(gene_symbol, TPM...25:TPM...46) %>%
  gather(key = "TPM_number", value = "value", 2:23, -gene_symbol) %>%
  spread(key = gene_symbol, value = value)

# generate an empty vector for storing the p values from the ttest and a vector with the names of the genes
p_value <- c()
gene_name <- colnames(H3K27M)

# use a for loop to perform a ttest for each gene and store the p value in the empty vector
for (i in 2:ncol(H3K27M)) {
  p_value[i] <- t.test(H3K27M[,i], WT[,i])$p.value
}

# generate a new dataframe that combines gene name and p value
ttest_results <- data.frame(gene_name, p_value)

# remove the first row of this data frame
ttest_results_final <- ttest_results[-1,]



#print out the t-test results

knitr::kable(ttest_results_final, format = "markdown", col.names = c("Gene", "p-value"), align = "c")
```

|    |   Gene   |  p-value  |
| :- | :------: | :-------: |
| 2  |  ADARB2  | 0.0000020 |
| 3  |   ALG5   | 0.0000817 |
| 4  |   ATRX   | 0.3072447 |
| 5  |   BRD2   | 0.1385084 |
| 6  |  BRINP3  | 0.0000306 |
| 7  |  CDKN2A  | 0.0430789 |
| 8  |  DCHS2   | 0.0000023 |
| 9  |  DIS3L2  | 0.3635906 |
| 10 |   EGFR   | 0.1642918 |
| 11 |   EZH2   | 0.9641300 |
| 12 |  FOXD1   | 0.0001875 |
| 13 |   GLB1   | 0.0003784 |
| 14 |   IDH1   | 0.1403354 |
| 15 |   IRX1   | 0.0000070 |
| 16 |   MGMT   | 0.3558852 |
| 17 |   MICA   | 0.0288078 |
| 18 |   OSR1   | 0.0000013 |
| 19 | PCDHGA11 | 0.2512366 |
| 20 |  PDGFRA  | 0.7507788 |
| 21 |  PIEZO1  | 0.0843526 |
| 22 |  PIK3CA  | 0.7354268 |
| 23 |   PTEN   | 0.0235358 |
| 24 |   RB1    | 0.1202909 |
| 25 |  STK38   | 0.0622940 |
| 26 |   TERT   | 0.6531105 |
| 27 |  TOB2P1  | 0.0001200 |
| 28 |   TP53   | 0.8962146 |
| 29 |  TRIM4   | 0.0003872 |
| 30 |  ZFPM2   | 0.0001880 |

Now create a graphing function to create boxplots to visualize the
results. Plot expression on the y-axis. The graph should look like [this
example](IRX1_example.png)

``` r
#to work in the tidyverse, it will be easier to make tidy the dataframe first



#create a graphing function

#then use a for loop combined with the graphing function to make a graph for all your genes of interest 
```

``` r
sessionInfo()
```

    ## R version 4.0.1 (2020-06-06)
    ## Platform: x86_64-apple-darwin17.0 (64-bit)
    ## Running under: macOS Catalina 10.15.6
    ## 
    ## Matrix products: default
    ## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
    ## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
    ## 
    ## locale:
    ## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
    ## 
    ## attached base packages:
    ## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
    ## [8] methods   base     
    ## 
    ## other attached packages:
    ##  [1] AnnotationFilter_1.12.0 GenomicFeatures_1.40.1  AnnotationDbi_1.50.3   
    ##  [4] Biobase_2.48.0          GenomicRanges_1.40.0    GenomeInfoDb_1.24.2    
    ##  [7] IRanges_2.22.2          S4Vectors_0.26.1        BiocGenerics_0.34.0    
    ## [10] knitr_1.29              forcats_0.5.0           stringr_1.4.0          
    ## [13] dplyr_1.0.0             purrr_0.3.4             readr_1.3.1            
    ## [16] tidyr_1.1.0             tibble_3.0.1            ggplot2_3.3.2          
    ## [19] tidyverse_1.3.0        
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] nlme_3.1-148                ProtGenerics_1.20.0        
    ##  [3] matrixStats_0.56.0          bitops_1.0-6               
    ##  [5] fs_1.4.2                    lubridate_1.7.9            
    ##  [7] bit64_4.0.2                 progress_1.2.2             
    ##  [9] httr_1.4.1                  tools_4.0.1                
    ## [11] backports_1.1.8             R6_2.4.1                   
    ## [13] lazyeval_0.2.2              DBI_1.1.0                  
    ## [15] colorspace_1.4-1            withr_2.2.0                
    ## [17] tidyselect_1.1.0            prettyunits_1.1.1          
    ## [19] bit_4.0.4                   curl_4.3                   
    ## [21] compiler_4.0.1              cli_2.0.2                  
    ## [23] rvest_0.3.5                 xml2_1.3.2                 
    ## [25] DelayedArray_0.14.1         rtracklayer_1.48.0         
    ## [27] scales_1.1.1                askpass_1.1                
    ## [29] rappdirs_0.3.1              Rsamtools_2.4.0            
    ## [31] digest_0.6.25               rmarkdown_2.3              
    ## [33] XVector_0.28.0              pkgconfig_2.0.3            
    ## [35] htmltools_0.5.0             highr_0.8                  
    ## [37] dbplyr_1.4.4                rlang_0.4.6                
    ## [39] readxl_1.3.1                rstudioapi_0.11            
    ## [41] RSQLite_2.2.0               generics_0.0.2             
    ## [43] jsonlite_1.7.0              BiocParallel_1.22.0        
    ## [45] RCurl_1.98-1.2              magrittr_1.5               
    ## [47] GenomeInfoDbData_1.2.3      Matrix_1.2-18              
    ## [49] Rcpp_1.0.4.6                munsell_0.5.0              
    ## [51] fansi_0.4.1                 lifecycle_0.2.0            
    ## [53] stringi_1.4.6               yaml_2.2.1                 
    ## [55] SummarizedExperiment_1.18.2 zlibbioc_1.34.0            
    ## [57] BiocFileCache_1.12.1        grid_4.0.1                 
    ## [59] blob_1.2.1                  crayon_1.3.4               
    ## [61] lattice_0.20-41             Biostrings_2.56.0          
    ## [63] haven_2.3.1                 hms_0.5.3                  
    ## [65] pillar_1.4.4                biomaRt_2.44.1             
    ## [67] reprex_0.3.0                XML_3.99-0.5               
    ## [69] glue_1.4.1                  evaluate_0.14              
    ## [71] modelr_0.1.8                vctrs_0.3.1                
    ## [73] cellranger_1.1.0            gtable_0.3.0               
    ## [75] openssl_1.4.2               assertthat_0.2.1           
    ## [77] xfun_0.15                   broom_0.5.6                
    ## [79] GenomicAlignments_1.24.0    memoise_1.1.0              
    ## [81] ellipsis_0.3.1
