---
title: "HW06"
output: github_document
---

Steps: 

  1. Read in the 45 processed RNA-Seq data found in "./RNA_Seq_processed"
  2. Map gene.ids to gene.symbols (which I have for you)
  3. For 27 genes of interest AND your favorite gene, perform a t-test to see if it is differentially expressed between the WT vs H3K27M samples
  4. Create a graphing function and then create a boxplot that graphs expression between the two groups
  
  
#### Code

```{r read in all 45 files as lists}
#create list of files contained by each directory 
H3K27M_list <- list.files("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/H3K27M")
WT_list <- list.files("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/WT")
```

```{r combine all mutated data}
library(readr)

#create empty dataset that loop results will go into 
H3K27M_df <- data.frame()
WT_df <- data.frame()

setwd("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/H3K27M")
#loop to read in mutated data
for (file in 1:length(H3K27M_list)) {
  H3K27M_data <- read_tsv(H3K27M_list[file])
  H3K27M_df <- rbind(H3K27M_df, H3K27M_data)
}
```

```{r combine WT data}
setwd("/Users/JuliaShangguan/Desktop/R coding class/HW06/RNA_Seq_processed/WT")
#loop to read in WT data
for (file in 1:length(WT_list)) {
  WT_data <- read_tsv(WT_list[file])
    WT_df <- rbind(WT_df, WT_data)
}
```

```{r specify whether gene is WT or mutated}
library(tidyverse)
library(magrittr)
library(dplyr)

#add type column to distinguish WT from mutated
WT_df <- WT_df %>% mutate("type" = "WT")
H3K27M_df <- H3K27M_df %>% mutate("type" = "H3K27_Mutated")
```

```{r}
#join into single dataframe for subsequent analysis
#my data is in long format, which is not what Robert's code was intended for, so see note below on some of Robert's code I had to omit 
GBM.transcripts <- WT_df %>%
full_join(H3K27M_df) %>%
  dplyr::rename("gene_id"="Name")

```

Using Robert's code

```{r map gene.ids to gene.symbols using ensembldb}

#install.packages("BiocManager")
#BiocManager::install("ensembldb")
#BiocManager::install("EnsDb.Hsapiens.v75")
library(ensembldb)
library(EnsDb.Hsapiens.v75)


ens.GBM.transcripts <- GBM.transcripts %>% 
mutate(gene_id = gsub(pattern = "\\..*", "", .$gene_id))

map <- ensembldb::select(EnsDb.Hsapiens.v75, keys = ens.GBM.transcripts$gene_id,
                         keytype = "TXID", columns = c("SYMBOL", "TXID"))

ens.mapped_GBM <- left_join(ens.GBM.transcripts, map, by = c("gene_id" = "TXID")) %>% 
  dplyr::select(-1) %>% 
  dplyr::select(gene_symbol = SYMBOL, everything())

#Because I combined my data long-wise, many values would be removed with the following code (because my gene names show up more than once). Robert told me that it's okay to not use the following functions and still be able to do my analysis. 

#ens.mapped_GBM <- ens.mapped_GBM[!duplicated(ens.mapped_GBM$gene_symbol),] #remove duplicated gene symbols
  #these are removed instead of averaged because they simply do not correlate particularly well. 

ens.mapped_GBM <- ens.mapped_GBM[!is.na(ens.mapped_GBM$gene_symbol),] #remove NA values

```

Do the t-test and make a table of the t-test results!

```{r t.test for difference between a bunch of genes}

#detach functions gave me an error, so I just specified which library some of my functions are from  
  #detach(package:EnsDb.Hsapiens.v75, unload = TRUE)
  #detach(package:ensembldb, unload = TRUE)
library(broom)

#added CDC42 as my own gene of interest
genes_of_interest <- c("IRX1", "OSR1", "DCHS2", "BRINP3", "TOB2P1", "FOXD1", "ZFPM2", "GLB1", "ALG5", "TRIM4", "ADARB2", "PCDHGA11", "IDH1", "EGFR", "MGMT", "TERT", "PTEN", "TP53", "RB1", "PTEN", "ATRX", "PDGFRA", "PIK3CA", "MICA", "CDKN2A", "EZH2", "BRD2", "CDC42")


GBM.genes.of.interest <- dplyr::filter(ens.mapped_GBM, gene_symbol %in% genes_of_interest)

#save table so I don't have to rerun all my previous code to do my analysis
#write_csv(GBM.genes.of.interest, "GBM.genes.of.interest.csv")

GBM.genes.of.interest <- read.csv("GBM.genes.of.interest.csv")


#creating function to filter by gene and then group by type
grouped_genes <- function(x){
  GBM.genes.of.interest %>%  
  dplyr::filter(gene_symbol == x) %>%
    group_by(type)
}

Group_genes <- lapply(genes_of_interest, grouped_genes)

#creating function to conduct two-sample t-test
ttest <- function(x){
  t.test(TPM~type, data = x) }
  
ttest_values <- lapply(Group_genes,ttest)

#relabel so it is clear what gene the statistics belong to
for (x in 1:length(ttest_values))  {
ttest_values[[x]]$data.name <-genes_of_interest[[x]]
 }

#print out the t-test results
print(ttest_values)

```


Now create a graphing function to create boxplots to visualize the results. Plot expression on the y-axis. The graph should look like [this example](IRX1_example.png)

```{r create a graphing function}

#selecting columns of interest to make new dataframe
trimmed_df <- GBM.genes.of.interest %>%  
  dplyr::select(gene_symbol, type, TPM)

#create graphing function
graph_tpm <- function(df) {
      plot <- ggplot(subset(trimmed_df, trimmed_df$gene_symbol == genes_of_interest[i]),
              aes(type, TPM, fill=type)) +
      geom_boxplot() +
      theme_bw()+
      scale_x_discrete(name = genes_of_interest[i]) +
      scale_y_continuous(name = "Expression_(TPM)") +
      ggtitle(paste(genes_of_interest[i], "Expression in GBM models \nby H3K27 Mutated or WT Status")) +
      theme(plot.title = element_text(hjust = 0.5)) +
      theme(legend.position = "none") 
  print(plot)
}

#using for loop with the graphing function
for(i in 1: length(genes_of_interest)) {
  graph_tpm(trimmed_df) 
}

```


```{r reproducibility}
sessionInfo()
```

